test/flush_worker_test.cc

```diff
@@ -0,0 +1,161 @@
+      auto guard = std::make_unique<Envoy::Thread::LockGuard>(lock_);
+      stopped_ = true;
+    }));
+  }
+
+  // Set up expected behaviors when run() is called on dispatcher_.
+  void expectDispatcherRun() {
+    EXPECT_CALL(*dispatcher_, run(_))
+        .WillOnce(Invoke([&](Envoy::Event::DispatcherImpl::RunType type) {
+          // The first dispatcher run is in WorkerImpl::start().
+          ASSERT_EQ(Envoy::Event::DispatcherImpl::RunType::NonBlock, type);
+        }))
+        .WillOnce(Invoke([&](Envoy::Event::DispatcherImpl::RunType type) {
+          // The second dispatcher run is in FlushWorkerImpl::work().
+          ASSERT_EQ(Envoy::Event::DispatcherImpl::RunType::RunUntilExit, type);
+          simulateTimerLoop();
+        }));
+  }
+
+  // Simulate the periodical timer which keeps running until dispatcher->exit()
+  // is called from another thread. The std::promise signal_dispatcher_to_exit_
+  // is used here to guarantee the while loop runs for at least 100 times before dispatcher->exit()
+  // is called.
+  void simulateTimerLoop() {
+    int i = 0;
+    while (true) {
+      i++;
+      // At least run the while loop for 100 times.
+      if (i == 100) {
+        signal_dispatcher_to_exit_.set_value();
+      }
+      auto guard = std::make_unique<Envoy::Thread::LockGuard>(lock_);
+      if (stopped_) {
+        break;
+      } else if (timer_set_) {
+        timer_set_ = false;
+        timer_cb_();
+      }
+    }
+  }
+
+  NiceMock<Envoy::Api::MockApi> mock_api_;
+  Envoy::Thread::ThreadFactory& thread_factory_;
+  Envoy::Stats::IsolatedStoreImpl store_;
+  NiceMock<Envoy::ThreadLocal::MockInstance> tls_;
+  NiceMock<Envoy::Event::MockDispatcher>* dispatcher_ = nullptr; // not owned
+  std::unique_ptr<Envoy::Runtime::ScopedLoaderSingleton> loader_;
+
+  NiceMock<Envoy::Event::MockTimer>* timer_; // not owned
+  Envoy::Event::TimerCb timer_cb_;
+  bool timer_set_{};
+  bool stopped_{false};
+  // Protect stopped_.
+  Envoy::Thread::MutexBasicLockable lock_;
+  std::promise<void> signal_dispatcher_to_exit_;
+
+  Envoy::Stats::MockSink* sink_ = nullptr; // owned by stats_sinks_
+  std::list<std::unique_ptr<Envoy::Stats::Sink>> stats_sinks_;
+};
+
+// Verify stats are flushed periodically until dispatcher->exit() is called from
+// another thread.
+TEST_F(FlushWorkerTest, WorkerFlushStatsPeriodically) {
+  std::chrono::milliseconds stats_flush_interval{10};
+  setupDispatcherTimerEmulation();
+
+  FlushWorkerImpl worker(mock_api_, tls_, store_, stats_sinks_, stats_flush_interval);
+
+  std::thread thread = std::thread([&worker, this] {
+    // Wait for the while loop run at least 100 times in simulateTimerLoop().
+    signal_dispatcher_to_exit_.get_future().wait();
+    worker.exitDispatcher();
+  });
+
+  expectDispatcherRun();
+  // Check flush() is called at least 100 times in simulateTimerLoop().
+  EXPECT_CALL(*sink_, flush(_)).Times(testing::AtLeast(100));
```

Comment:
You know this better to me, so I'll defer to you.  Do you have any fears that this might be brittle or flaky?   Is this running 100 times a pretty safe assumption?  Or alternatively, if it doesn't run over 100 times, is that definitely indicative of a problem?

Reply:
reading this, I wanted to float a suggestion: in another part of this source file there's:

```c++
      if (i == 100) {
        signal_dispatcher_to_exit_.set_value();
      }
```

I feel it would help readability if that was a nicely named constant. 
Also, while taking another look at this, removing the simulated_time_system header / dependency might be 
helpful, because I don't think simulated time is actually used, correct?

Reply:
This comment doesn't seem addressed yet.

Reply:
Thanks Nate and Otto.

I think the test is not going to be flaky since the loop is guaranteed to run at least 100 times. (@oschaaf to correct me if I'm wrong.)
This is because 
 - `signal_dispatcher_to_exit_.get_future().wait()` will make the call `worker.exitDispatcher()` wait after the loop run at least 100 times
- the loop will keep running until `worker.exitDispatcher()` is called, so the loop is guaranteed to run >= 100 times.

Also defined a const int kNumTimerLoops to replace 100 and removed simulated_time_system header.

Reply:
@qqustc I had some hands on with the test code, to see if this could be simplified, and ended up with a small change I'd like to float for consideration: https://github.com/qqustc/nighthawk/compare/pr1...oschaaf:pr1#diff-e0be675b84ffe6e6fd41947249a3bafe
(please ignore all other changes, only the changes to `flush_worker_test.cc` are relevant, the rest was induced by merging master).

Tangentially related, there might be an opportunity for another potential enhancement, perhaps best done in an follow up iteration if consensus is that this is desirable: Maybe we can refactor `SequencerTestWithTimerEmulation` to be more easily shareable, so we consolidate some of the complexity that comes with that into a single place. An aside to that would be that we will have simulated time flowing as well here, which can be used for something that might be useful: testing the number of periodic flushes based on the actual passage of (simulated) time.

Reply:
Thank you Otto for the simplification!

For refactoring the the complexity of SequencerTest and FlushWorkerTest into a single place, since the 2 tests are still quite different even though they both simulate timer, I'm not sure the benefit is worth the hassle of the refactoring work.